filtering/checker_test.go

```diff
@@ -0,0 +1,398 @@
+
+import (
+	"testing"
+
+	syntaxv1 "go.einride.tech/aip/examples/proto/gen/einride/example/syntax/v1"
+	"gotest.tools/v3/assert"
+)
+
+func TestChecker(t *testing.T) {
+	t.Parallel()
+	for _, tt := range []struct {
+		filter        string
+		declarations  []DeclarationOption
+		errorContains string
+	}{
+		{
+			filter: "New York Giants",
+			declarations: []DeclarationOption{
+				DeclareIdent("New", TypeBool),
+				DeclareIdent("York", TypeBool),
+				DeclareIdent("Giants", TypeBool),
+			},
+			errorContains: "undeclared function 'FUZZY'",
+		},
+
+		{
+			filter: "New York Giants OR Yankees",
+			declarations: []DeclarationOption{
+				DeclareStandardFunctions(),
+				DeclareIdent("New", TypeBool),
+				DeclareIdent("York", TypeBool),
+				DeclareIdent("Giants", TypeBool),
+				DeclareIdent("Yankees", TypeBool),
+			},
+			errorContains: "undeclared function 'FUZZY'",
+		},
+
+		{
+			filter: "New York (Giants OR Yankees)",
+			declarations: []DeclarationOption{
+				DeclareStandardFunctions(),
+				DeclareIdent("New", TypeBool),
+				DeclareIdent("York", TypeBool),
+				DeclareIdent("Giants", TypeBool),
+				DeclareIdent("Yankees", TypeBool),
+			},
+			errorContains: "undeclared function 'FUZZY'",
+		},
+
+		{
+			filter: "a b AND c AND d",
+			declarations: []DeclarationOption{
+				DeclareStandardFunctions(),
+				DeclareIdent("a", TypeBool),
+				DeclareIdent("b", TypeBool),
+				DeclareIdent("c", TypeBool),
+				DeclareIdent("d", TypeBool),
+			},
+			errorContains: "undeclared function 'FUZZY'",
+		},
+
+		{
+			filter: "a",
+			declarations: []DeclarationOption{
+				DeclareIdent("a", TypeBool),
+			},
+		},
+
+		{
+			filter: "(a b) AND c AND d",
+			declarations: []DeclarationOption{
+				DeclareStandardFunctions(),
+				DeclareIdent("a", TypeBool),
+				DeclareIdent("b", TypeBool),
+				DeclareIdent("c", TypeBool),
+				DeclareIdent("d", TypeBool),
+			},
+			errorContains: "undeclared function 'FUZZY'",
+		},
+
+		{
+			filter: `author = "Karin Boye" AND NOT read`,
+			declarations: []DeclarationOption{
+				DeclareStandardFunctions(),
+				DeclareIdent("author", TypeString),
+				DeclareIdent("read", TypeBool),
+			},
+		},
+
+		{
+			filter: "a < 10 OR a >= 100",
+			declarations: []DeclarationOption{
+				DeclareStandardFunctions(),
+				DeclareIdent("a", TypeInt),
+			},
+		},
+
+		{
+			filter: "NOT (a OR b)",
+			declarations: []DeclarationOption{
+				DeclareStandardFunctions(),
+				DeclareIdent("a", TypeBool),
+				DeclareIdent("b", TypeBool),
+			},
+		},
+
+		{
+			filter: `-file:".java"`,
+			declarations: []DeclarationOption{
+				DeclareStandardFunctions(),
+				DeclareIdent("file", TypeString),
+			},
+		},
+
+		{
+			filter:        "-30",
+			errorContains: "non-bool result type",
+		},
+
+		{
+			filter: "package=com.google",
+			declarations: []DeclarationOption{
+				DeclareStandardFunctions(),
+				DeclareIdent("package", TypeString),
+				DeclareIdent("com", TypeMap(TypeString, TypeString)),
```

Comment:
It comes from the EBNF - since we don't support records yet, I declared it as a map to make the type-check pass.

Reply:
But isn't it supposed to be interpreted as a string?

Reply:
Since we do the "." operator on it, we need to interpret it as some type of "package"-like type that we can do the select operator on. It's probably intended to be a record, but I made it a map<string, string> since that also works.